pkg/image/image.go

```diff
@@ -26,31 +26,57 @@ type Image struct {
 	FileCatalog FileCatalog
 }
 
-// NewImage provides a new, unread image object.
-func NewImage(image v1.Image) *Image {
-	return &Image{
-		content:     image,
-		FileCatalog: NewFileCatalog(),
+type AdditionalMetadata func(*Image) error
```

Comment:
if it helps, what sold me is that it seems like we may be adding an arbitrary amount of optional data with unknown processing needs --didn't feel right for the function signature to be at the mercy of such changes, and a static config didn't seem right either.

Reply:
(this is a comment/exploration more on the pattern than this particular code or PR!)

@wagoodman I think the interesting here is that, as we're constructing the image, what we're in need of is _data_ (like a tag or a digest), but what `NewImage` is asking for is _functions_. This is addressing a need, _and then some_.

Like we mentioned, because Go doesn't support multiple function signatures, it does get weird thinking about how to add data to an object, when you might not have all the data pieces every time you go to construct the object.

Outside of passing in a struct with properties that can be `nil`, I've also seen the notion of "fluent code" help out here, which is sort of what this pattern is approaching.

Something like:

```go
image.NewImage(img).WithTags(tags).WithManifest(theManifest.raw)
```

This ends up looking very similar to the current implementation, and it solves the same problem. With this fluent approach, the consumer doesn't need to take care of providing function implementations. Given that the set of kinds of extra data (e.g. tags, manifest, digest) is finite and known, the consumer can pass just the _data_ that they want to include in the constructed object, as opposed to a _function_ that is needed to help with the actual construction itself. Construction-functionality is performed strictly by the object in such a way that's hidden (and inaccessible) to the consumer.

Another benefit of this approach is that it forces the scope of potential error paths to be more confined. Since the chaining approach implies a single object is being returned from each call, errors must be dealt with upstream. Values whose processing might return an error (e.g. tags) can be handled ahead of construction, and values that are available without any potential errors can be used as is (as opposed to being returned by a delegate function that has an `error` return type but never actually returns an error).

One other thought I had was that — for cases when the struct in question is not overly complex or nested (which might not apply here) — it could be that Go's struct literal "API" is good enough, and there's not much added value in a constructor of any kind. Since struct literals natively handle the zero value for fields not specified, this allows for a Go-native solution for customizing the resulting object based on what data is available.

Reply:
re: method chaining on objects: I'm aware of this approach and think that it is a great pattern. I don't think it's a good fit in this case in this PR (I realize you noted this comment is not on the PR, but it makes for a useful reflection). 

providing data conditionally looks pretty similar (as you mention):
```golang
// create the image...
i := image.NewImage(img)

if <somecondition> {
  i = i.WithTags(tags)
}

if <somecondition> {
  i = i.WithManifest(theManifest.raw)
}

return i

```

or if we're saying that we're allowing the calls to mutate the reference
```golang
// create the image...
i := image.NewImage(img)

if <somecondition> {
   i.WithTags(tags)
}

if <somecondition> {
   i.WithManifest(theManifest.raw)
}

return i

```

the API doesn't suggest that all of this information should be provided at creation so you can do things like:
```golang
// create the image...
i := image.NewImage(img)

if <somecondition> {
  i = i.WithTags(tags)
}

// use the image
i.Read() 

// ... add more data to the image? this seems wrong for the API to suggest we should do this <-----
if <somecondition> {
  i = i.WithManifest(theManifest.raw)
}

return i

```

> With this fluent approach, the consumer doesn't need to take care of providing function implementations

With the current approach the consumer doesn't need to implement functions at all, yes? Only provide already implemented functions. If they choose to bring their own that does it's own implementation (which is not necessary), this is ok! This is probably better than forcing the user to mutate an object after creation out of necessity:

```
i = image.New(...)

i.Metadata.Thing = "custom value"
```

Dunno why someone would want to, but we don't prevent it either.

> benefit of this approach is that it forces the scope of potential error paths to be more confined

This is now less confined rather than more confined. Let's look at `WithTags(tags []string) *Image`, there is processing that can result in an error, you mention:

> Values whose processing might return an error (e.g. tags) can be handled ahead of construction

We are just moving error processing out of the confined single function where we are constructing the object and moving it out to the caller. Why should the caller need to individually handle construction of non-literal elements?


...

Overall a config struct would be a better, simpler way than with supporting chaining. There is no inherent problem with this approach.

Reply:
>the API doesn't suggest that all of this information should be provided at creation so you can do things ...

If taking my one-liner literally, yes this is true, and this conclusion makes sense. But usually this kind of pattern demarcates the construction with a `.Build` method or something similar, if it's for an object's construction. I think it would've been clearer had I rounded out the snippet I included.

>With the current approach the consumer doesn't need to implement functions at all, yes? Only provide already implemented functions.

I'm not sure where the confusion was here, but yes, that's what I'm saying. The consumer needs to provide _functions_... as a means of providing _data_ (and specifically, not "to-be-computed" data, but data whose state is already known). This is one significant point about the pattern that's currently implemented. I'd like to better understand if the intent is for the developer that consumes these functions to understand that they're providing functions that successively mutate an object. If so, I think we need to better understand how intuitive this is for the other "users" of the code base (i.e. the other members of the dev team). If not, this detracts from maintainability because it's not obvious what the impact of a change would be.

Regarding error path confinement: It looks like right now, the function signature includes an error return value (looking at `type AdditionalMetadata func(*Image) error`). The `WithManifest` function doesn't ever return an error, but its caller must check for an error regardless. The `WithManifestDigest` function doesn't ever return an error, but its caller must check for an error regardless. The `WithTags` function _can_ return an error, which means the other two functions must conform to the notion of returning an error-type value, which means these functions' callers must always check for the possibility of a non-nil error. (Side note: it looks like the `WithTags` function needs to be able to return an error only because it's receiving `string` tags as an argument and then calls `name.NewTag` to transform those values. This call to `NewTag` could've happened upstream, such as in the `manifest` struct's `tags` method, in which case `WithTags` could've just received `name.Tag`s.)

The net result of all this ^ is that `NewImage` needs to conditionally return an error, too, because the `AdditionalMetadata` function can return an error, and again, that's just because of one out of three functions that currently can return an error (but don't necessarily need to). This is what I meant regarding the error confinement observation.

Regarding this conclusion:

>Ultimately having a single point where we create the object and allow for input of optional data I feel is the ideal goal (a choke point for object generation).

I see the appeal. From what I heard earlier today, you're interested in deduplicating code, such that future changes needn't be made in multiple spots for a single effect. That resonates.

I think the heart of the problem here is that it's arbitrary when the image object is said to be done being "constructed". No matter the form factor, if we're not specifying a struct literal, that means we're iteratively amended the returned object as we add more field data. We could do this amending during in the consuming function, or in the constructor. If we do it in the constructor, we can take functions (the current approach) or we can take data with a finalization call (the fluent approach).

Ultimately, I think it's difficult to make a case for one of these approaches being "more pure" than the other, since all approaches require handling the conditionality of _having_ vs. _not having_ any given piece of data.

Which — _to me_ — leaves the consumer itself as the worthwhile factor to consider. It seems to me that creating a slice of mutating functions, then unpacking that slice inside a variadic parameter, is less straightforward than providing just the data _(digest/manifest/tags/whatever)_ that was discovered.